_posts/2024-01-21-comparing-cas-and-karpenter.markdown

```diff
@@ -0,0 +1,273 @@
+availability and pricing. By contrast, to emulate this behavior using the CAS
+with Cluster API would require the user to be updating the
+`InfrastructureMachineTemplate` resources in their cluster as well as updating
+the associated scalable types on a continual basis.
+
+## When to activate, pending or unscheduled?
+
+Another subtle point of difference that flows from the notion of scaling
+versus provisioning is the conditions under which these applications will jump
+into action.
+
+When there are pending pods in a cluster, this means that there
+is a node available which could receive the pod but that it currently does not
+have allocatable capacity. In these cases, when configured, the CAS and
+Karpenter both have the ability to  add more nodes of the type that will accept
+the workload.
+
+When there are unschedulable pods in a cluster, this means that there are no
+nodes that can satisfy the requirements of the pod. In these cases the CAS
+will only act if it has a node group which could possibly make a node to
+schedule that pod. This scenario can happen when the CAS is configured to
+have node groups of size zero and thus there are no nodes in the cluster which
+could schedule the pod, but the CAS knows how to make that type of node.
+
+Karpenter, by contrast, is configured by the user to instruct the provisioning
+of nodes based on pod constraints. In the case of an unschedulable pod,
+Karpenter will refer to its provisioners, and will then request several
+instance types which could satisfy the pod requirements. Karpenter can then
+choose the best (by cost, resource, availability, etc) instance type to create
+as the pod is being requested.
+
+This is not to say that the CAS won't also look at several instance types when
+deciding which to make, but we need to look at the details a little closer to
+understand that choice. In Karpenter, having the ability to calculate instance
+types based on the workload constraints means that it can request a broad
+range of instances (limited by user configuration) from the infrastructure. By
+contrast, CAS can be configured with many instance types and when presented
+with a decision about which type of instance to request for a specific workload
+it will use the user's preference (up to and including [custom code][cas-grpc])
+to make that choice. The CAS is limited by its node group configurations,
+which are different on each platform, and may or may not support dynamic
+instance discovery and full resource and cost data.
+
+In both cases, CAS and Karpenter, the user has control over the options for
+how nodes are chosen, with preferences and priorities across several ranges
+such as pricing and resource consumption.
+
+## Disrupting behavior and bin packing
+
+Another topic that comes up frequently when talking about Karpenter is
+[bin packing][binpack]. Bin packing in this context refers to the
+algorithms that a program uses to fit items into a set of containers (bins)
```

Comment:
Nit, using containers in this context may cause confusion with the concept of podman/docker containers, perhaps choose a synonym that isn't so close to an adjacent topic

eg number of groups, number of slots

Reply:
i had a similar thought, which is why i put "(bins)" in there, but i should change it.